src/ucp/api/ucpx.h

```diff
@@ -19,7 +19,63 @@
 
 BEGIN_C_DECLS
 
+enum ucp_am_rendezvous_params_field {
+  UCP_AM_RENDEZVOUS_FIELD_IOVEC_SIZE = UCS_BIT(0)  /* Max size of the iovec returned by the initial AM */
+};
 
+typedef struct ucp_am_rendezvous_params {
+  uint64_t field_mask ;
+  size_t iovec_size ;
+} ucp_am_rendezvous_params_t ;
+
+
+typedef void (*ucp_am_data_function_t)(void *target, void *source, size_t bytes, void *cookie) ;
+typedef ucs_status_t (*ucp_am_local_function_t)(void *arg, void *cookie, ucp_dt_iov_t *iovec, size_t iovec_length) ;
+
+typedef struct ucp_am_rendezvous_recv {
+    ucp_am_local_function_t local_fn ;
+    void * cookie ;
+    ucp_am_data_function_t data_fn ;
+    void *data_cookie ;
+    size_t iovec_max_length ;
+    size_t iovec_length ;
+    ucp_dt_iov_t iovec[1] ;
+  } ucp_am_rendezvous_recv_t ;
+
+typedef ucs_status_t (*ucp_am_rendezvous_callback_t)(void *arg, void *data, size_t length,
+                                          ucp_ep_h reply_ep, unsigned flags, size_t remaining_length, ucp_am_rendezvous_recv_t *recv);
```

Comment:
is `recv` an `out` parameter? or just partially? What is the flow if user doesn't know receive buffer details when rendezvous callback is invoked? Will the AM header be queued or reply will be sent requesting eager delivery of the message?

Reply:
`recv` is mostly out; the `iovec_max_length` field is in, and the rest of the fields are out.
The design is for the case where the server application does know the receive buffer when the rendezvous callback is invoked; but there is enough information passed so that the server function could call `malloc` to allocate a buffer if wanted. The design intent is for the rdma transfer to start as soon as the rendezvous callback returns; there is no facility for queueing or getting the message delivered eagerly.